hazelcast/src/test/java/com/hazelcast/jet/JobListenerTest.java

```diff
@@ -0,0 +1,274 @@
+                    assertNotNull(failure);
+                    assertTailEquals(listener.log,
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> FAILED (" + failure + ")");
+                });
+    }
+
+    @Test
+    public void testListener_restartOnException() {
+        testListener(new JobConfig().setAutoScaling(true),
+                TestSources.itemStream(1,
+                        (t, s) -> {
+                            if (s == 2) {
+                                throw new RestartableException();
+                            }
+                            return s;
+                        }),
+                (job, listener) -> {
+                    assertEqualsEventually(listener::runCount, 2);
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log.stream().map(SIMPLIFY).collect(toList()),
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> NOT_RUNNING (com.hazelcast.jet.RestartableException)",
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "User: RUNNING -> FAILED (CANCEL)");
+                });
+    }
+
+    @Test
+    public void testListener_suspendOnFailure() {
+        testListener(new JobConfig().setSuspendOnFailure(true),
+                TestSources.itemStream(1, (t, s) -> 1 / (2 - s)),
+                (job, listener) -> {
+                    assertJobStatusEventually(job, SUSPENDED);
+                    String failure = job.getSuspensionCause().errorCause().split("\n", 3)[1];
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log,
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> SUSPENDED (" + failure + ")",
+                            "User: SUSPENDED -> FAILED (CANCEL)");
+                });
+    }
+
+    @Test
+    public void testListenerDeregistration() {
+        testListener(TestSources.itemStream(1, (t, s) -> s),
+                (job, listener) -> {
+                    assertJobStatusEventually(job, RUNNING);
+                    instance.get().getJet().removeJobStatusListener(listener);
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log,
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING");
+                });
+    }
+
+    @Override
+    public void supportAfter() {
```

Comment:
Instead of disabling the superclass' after-code, make the test serial on line 59. This way the logs are also much easier to analyze.

Reply:
Switching to the serial test runner increased the total execution time significantly.

Reply:
In that case perhaps we can switch back to parallel test, and consider not subclassing the `SimpleTestInClusterSupport`. This base class is used a lot and I'm used to rely that it cleans all the distributed objects and cancel all jobs. By removing the _after_ code, you perhaps are removing more of its behavior than you're keeping.

Perhaps we can simply keep the test as is and mark it as NightlyTest, or switch back to a parallel test and not reuse the cluster. Or implement a reused cluster in this class. But in my experience, parallel tests mess up the logs, you don't know which log line is produced by which test, and if there's a random test failure, if you can't reproduce it locally, the logs from jenkins are almost useless to get a clue what has happened.

Reply:
Switching back to parallel test will make the logs useless as you pointed out. I think, it's enough to mark the test as `SlowTest`, so that it's skipped by the PR builder, but included in the code coverage.